### PR TITLE
prepare-root: Set up sysroot readonly in initramfs

### PR DESCRIPTION
**Pull request by cgwalters**
Let's ensure things are right from the start in the initramfs;
this closes off various race conditions.  Followup to
https://github.com/ostreedev/ostree/pull/2113/commits/35642259175973617da937f3cab6ce5f13c95077

Closes: https://github.com/ostreedev/ostree/issues/2115

### CODE REVIEW

**Comment by cgwalters** [comment]
This currently fails in FCOS since `igniton-files` can't run `useradd`; still debugging, the mounts all seem right though:

```
`-/sysroot                            /dev/vda4[/ostree/deploy/fedora-coreos/deploy/24baddf67040f815aa56043b20001aa94c80ca10a7f4b9441e8640562c968318.0]
                                                 xfs     rw,relatime,attr2,inode
  |-/sysroot/sysroot                  /dev/vda4  xfs     ro,relatime,attr2,inode
  |-/sysroot/etc                      /dev/vda4[/ostree/deploy/fedora-coreos/deploy/24baddf67040f815aa56043b20001aa94c80ca10a7f4b9441e8640562c968318.0/etc]
  |                                              xfs     rw,relatime,attr2,inode
  `-/sysroot/usr                      /dev/vda4[/ostree/deploy/fedora-coreos/deploy/24baddf67040f815aa56043b20001aa94c80ca10a7f4b9441e8640562c968318.0/usr]
                                                 xfs     ro,relatime,attr2,inode

```

**Comment by smuda** [comment]
What is the status of this issue? 

I'm running into what looks very much like [fedora-coreos issue #746](https://github.com/coreos/fedora-coreos-tracker/issues/746) on stable FCOS releases 34.20210529.3.0 and 34.20210821.3.0 but works on FCOS Testing release 34.20210904.2.0. Is it possible this has been merged to testing but now yet stable?

**Comment by jlebon** [comment]
> What is the status of this issue?
> 
> I'm running into what looks very much like [fedora-coreos issue #746](https://github.com/coreos/fedora-coreos-tracker/issues/746) on stable FCOS releases 34.20210529.3.0 and 34.20210821.3.0 but works on FCOS Testing release 34.20210904.2.0. Is it possible this has been merged to testing but not yet stable?

Interesting. It's not in any Fedora CoreOS release, so something else must be at play. The latest stable release from this week (34.20210904.3.0) is a promotion of 34.20210904.2.0, so I'd give that a try. If it still doesn't work, please file an issue on the FCOS tracker.

File: src/switchroot/ostree-prepare-root.c [hunk 1]
```diff
@@ -320,12 +318,45 @@ main(int argc, char *argv[])
...
+  bool mount_var = true;
+#endif
+  /* File in /run can override the default behaviour so that we definitely mount /var */
+  if (lstat (INITRAMFS_MOUNT_VAR, &stbuf) == 0)
+    mount_var = true;
+
+  /* Link to the deployment's /var */
+  if (mount_var)
+    {
+      if (snprintf (srcpath, sizeof(srcpath), "%s/../../var", deploy_path) < 0)
```
[the remark this answers is not in the document]
Comment (lucab) [reply src/switchroot/ostree-prepare-root.c]: It was semi-intentional, as I was trying to make it work initially by using absolute paths for my own sanity/readability.
It doesn't have any visible side-effects, as the kernel performs path normalization on mount anyway.
In order to reduce the scope of changes in this PR, I've kept using relative paths everywhere.

**Comment by lucab** [comment]
I've drained all the drive-by changes into separate PRs (all already merged: https://github.com/ostreedev/ostree/pull/2471, https://github.com/ostreedev/ostree/pull/2472, https://github.com/ostreedev/ostree/pull/2473, https://github.com/ostreedev/ostree/pull/2475), and adopted a few other suggestions from previous review.
Fully rebased on `main` now and ready to go.

**Comment by cgwalters** [comment]
> I've drained all the drive-by changes into separate PRs (all already merged: #2471, #2472, #2473, #2475), and adopted a few other suggestions from previous review.

Thanks so much for doing that!  In the end, I am *fairly* hopeful this change will sail through regression free, but if we discover a problem at least now it's easy to revert just this last bit while keeping all the cleanups, etc.

---

So, I just now discovered because I'm listed as the original author, I can't approve this PR.  Which is clearly a gap in the system.  Anyways, from my PoV feel free to approve the PR!

**Review by jlebon — APPROVED** [review]
Nice and clean. Thanks a lot for working on this!

**Pull request merged**